File: compress/utils/enc_map.m

function [R_Amax] = enc_map(Ctarget, space, comp_sel, C_flag, conv1_half)

filename = ['base/Net_def.txt'];
filename1 =['base/R_list.txt'];
filename2 =['base/A_norm.txt'];
filename3 =['base/W_norm.txt'];
filename4 =['base/C_norm.txt'];


delimiterIn = ' ';
Net = importdata(filename,delimiterIn);

num_fc = Net(:,7)-Net(:,8);
num_fc = sum(num_fc == 0)
if comp_sel == 2
    L = size(Net,1)
else
    L = size(Net,1)-num_fc
end

if comp_sel == 0
    Net = Net(1:L,:)
end

a = Net(:,8)./Net(:,2);
w = Net(:,7)./Net(:,2);

W_orig = sum(Net(:,7)); 
C_orig = sum(Net(:,8)); 
FC_cost = sum(Net(end-num_fc+1:end,7));


R1 = importdata(filename1,delimiterIn);
R1 = R1(:,1:L);
R = [zeros(1,L) ; R1 ];

A1 = importdata(filename2,delimiterIn);
A1 = A1(:,1:L);
A = [zeros(1,L) ; A1 ];

W1 = importdata(filename3,delimiterIn);
W1 = W1(:,1:L);
W = [zeros(1,L) ; W1 ];

C1 = importdata(filename4,delimiterIn);
C1 = C1(:,1:L);
C = [zeros(1,L) ; C1 ];

R_max = R(end,:);
R_norm = R(:,:)./R_max;
 

R_max = R(end,:);
R_norm = R(:,:)./R_max;
 
% Accuracy
A_norm = A(:,:)./A(end,:)
a_max = A(end,1);

% Weight
Wmax = W(end,:)
W_norm = W(:,:)./Wmax;

% FLOPs
Cmax = C(end,:)
C_norm = C(:,:)./Cmax;

n = size(R_norm,1);
m = size(R_norm,2);


%% Generation Cost Function - from A

a_norm = [0.1:0.01:1];
max_rank = [];
for i=1:size(A_norm,2)

    A_norm_tmp{i} = A_norm(:,i);
    R_norm_tmp{i} = R_norm(:,i);

    A_mask = A_norm_tmp{i} < 1;
    A_idx(i) = sum(A_mask)+1;

    A_norm_tmp{i} = A_norm_tmp{i}(1:A_idx(i));
    R_norm_tmp{i} = R_norm_tmp{i}(1:A_idx(i));

    [~, ind] = unique(A_norm_tmp{i});
    A_norm_tmp{i} = A_norm_tmp{i}(ind);
    R_norm_tmp{i} = R_norm_tmp{i}(ind);

    [~, ind] = unique(R_norm_tmp{i});
    A_norm_tmp{i} = A_norm_tmp{i}(ind);
    R_norm_tmp{i} = R_norm_tmp{i}(ind);


    diff_a = [];
    A_idx(i) = length(A_norm_tmp{i});
    for k=1:A_idx(i)-1
        diff_a(k) = A_norm_tmp{i}(k+1) - A_norm_tmp{i}(k);
    end
    diff_mask = diff_a<=0;
    diff_num = sum(diff_mask);
    tmp_a = A_norm_tmp{i}(2:end);
    tmp_a = tmp_a(~diff_mask); % +1 remove
    tmp_r = R_norm_tmp{i}(2:end);
    tmp_r = tmp_r(~diff_mask); % +1 remove
    A_norm_tmp{i} = [A_norm_tmp{i}(1);tmp_a];
    R_norm_tmp{i} = [R_norm_tmp{i}(1);tmp_r];

    A_idx(i) = A_idx(i) - diff_num;
    max_rank(i) = R_norm_tmp{i}(end);
end


for n=1:size(a_norm,2)
    for i=1:size(A_norm,2)
       
		x = A_norm_tmp{i};
		y = R_norm_tmp{i};

        pr_int(i) = round(interp1(x, y, a_norm(n),'pchip')*R(end,i));
    end
    r_min_vbmf(n,:) = pr_int;
end



%% Cost Calculation - from A
Test_R = r_min_vbmf;
Test_C = Test_R.*a';
Test_W = Test_R.*w';

if conv1_half == 1
    Test_W(:,1) = 0;
    Test_C(:,1) = 0;
end

W_acc = sum(Test_W,2)./W_orig;
C_acc = sum(Test_C,2)./C_orig;


%% Refinement of complexity 
if Ctarget > max(C_acc)
    for n=1:size(a_norm,2)
        for i=1:size(A_norm,2)
            R_norm_tmp{i}(end)=1;
            x = A_norm_tmp{i};
            y = R_norm_tmp{i};
            pr_int(i) = round(interp1(x, y, a_norm(n),'pchip')*R(end,i));
        end
        r_min_vbmf(n,:) = pr_int;
    end

    Test_R = r_min_vbmf;
    Test_C = Test_R.*a';
    Test_W = Test_R.*w';

    if conv1_half == 1
        Test_W(:,1) = 0;
        Test_C(:,1) = 0;
    end

    W_acc = sum(Test_W,2)./W_orig;
    C_acc = sum(Test_C,2)./C_orig;
end

% C_flag - 0 : Weight / 1 : FLOPs
if C_flag == 0
    C_acc = W_acc;
end

if conv1_half == 1
    if C_flag == 1
        Ctarget = Ctarget - Net(1,8)/Net(1,2)*round(Net(1,2)/2)/C_orig
    elseif C_flag == 0
        Ctarget = Ctarget - Net(1,7)/Net(1,2)*round(Net(1,2)/2)/W_orig
    end
end


%% Interpolate function
[~, ind] = unique(a_norm);
a_norm = a_norm(ind);
C_acc = C_acc(ind);

[~, ind] = unique(C_acc);
a_norm = a_norm(ind);
C_acc = C_acc(ind);

a_norm = [0 a_norm];
C_acc = [0; C_acc];

[C_acc_, W_acc_, Rmax] = Ro_calculation(Ctarget, space, C_acc, a_norm, A_norm, A_norm_tmp, R_norm_tmp, A_idx, R, a, w, conv1_half, C_orig, W_orig, C_flag)

C_acc_out(1) = C_acc_
W_acc_out(1) = W_acc_

[C_acc_, W_acc_, Rmin] = Ro_calculation(Ctarget, -space, C_acc, a_norm, A_norm, A_norm_tmp, R_norm_tmp, A_idx, R, a, w, conv1_half, C_orig, W_orig, C_flag)

C_acc_out(2) = C_acc_
W_acc_out(2) = W_acc_

if conv1_half == 1
    Rmax(1) = round(Net(1,2)/2);
    Rmin(1) = round(Net(1,2)/2);
    C_acc_out = C_acc_out + Net(1,8)/Net(1,2)*Rmax(1)/C_orig;
    W_acc_out = W_acc_out + Net(1,7)/Net(1,2)*Rmax(1)/W_orig;
end

Rmax
Rmin
C_acc_out
W_acc_out

% file write
Set = [Rmax, Rmin, C_acc_out, W_acc_out];
save(['tmp_/MATLAB_result.txt'], 'Set', '-ascii');

disp('~~~~~~~~~~~~~~~~~~~~~~')    
disp('~~ Boundary Gen End ~~ ');
disp('~~~~~~~~~~~~~~~~~~~~~~')    



end



function [C_acc_o, W_acc_o, Rmax] = Ro_calculation(Ctarget, space, C_acc, a_norm, A_norm, A_norm_tmp, R_norm_tmp, A_idx, R, a, w, conv1_half, C_orig, W_orig, C_flag)

    C_acc_ = C_acc;
    end_flag = 1
    delta = 0;
    k = 0;
    while(end_flag>0)
        Amax = interp1(C_acc_,a_norm,Ctarget+space+delta,'pchip'); 
        if Ctarget+space+delta > 1
            Amax = 1.0;
        end
        acc_min_idx = sum(a_norm<0.9);

        L = size(A_norm,2);
        for i=1:L    
            x = A_norm_tmp{i}(1:A_idx(i));
            y = R_norm_tmp{i}(1:A_idx(i));
            Rmax(i) = max(interp1(x, y, Amax,'pchip'),0); 
        end
        Rmax = round(Rmax.*R(end,1:L));
        Test_R = Rmax;
        W_acc = [];
        C_acc = [];

        Test_C = Test_R.*a';
        Test_W = Test_R.*w';


        if conv1_half == 1
            Test_W(1) = 0;
            Test_C(1) = 0;
        end


        W_acc = sum(Test_W)./W_orig;
        C_acc = sum(Test_C)./C_orig;

        if C_flag == 0
            C_acc = W_acc;
        end

        if C_acc <= Ctarget+space
            delta = delta+0.001;
        else 
            delta = delta-0.001;
        end
        k = k+ 1;

        if (C_acc < (Ctarget +space + 0.001)) && ((Ctarget +space - 0.001) < C_acc)
            flag_end = 0;
        end
        if k > 500
            flag_end = 0;
            break;
        end
    end
    C_acc_o = C_acc;
    W_acc_o = W_acc;
end

File: compress/utils/enc_map.py
import numpy as np
import copy
import json
import os
import os.path as osp
import sys
import time
import itertools
import google.protobuf as pb
import random

from argparse import ArgumentParser
from pprint import pprint

import subprocess
from scipy import interpolate
from scipy.interpolate import interp1d
from itertools import groupby

from basic import *


def make_unique_points(L, E_norm, R_norm):
    E_norm_tmp = []
    R_norm_tmp = []
    E_idx = []
    for i in range(L):
        #print '0',E_norm[:,i], R_norm[:,i]
        idx_ = np.sum(E_norm[:,i]<1)+1
        E_norm_ = E_norm[:idx_,i]
        #print '1',E_norm_
        R_norm_ = R_norm[:idx_,i]
    
        E_norm_, idx_ = np.unique(E_norm_,return_index=True)
        R_norm_ = R_norm_[idx_]
        #print '2',E_norm_
        R_norm_, idx_ = np.unique(R_norm_,return_index=True)
        E_norm_ = E_norm_[idx_]
        #E_norm_[-1] = 1
        #print '3',E_norm_

        diff_E = np.array([E_norm_[k+1]-E_norm_[k] for k in range(len(E_norm_)-1)])
        diff_mask = (diff_E <= 0)

        tmp_a = E_norm_[1:]
        tmp_a = tmp_a[~diff_mask]
        tmp_r = R_norm_[1:]
        tmp_r = tmp_r[~diff_mask]


        idx_ = len(E_norm_) - np.sum(diff_mask)
#        E_norm_ = [E_norm_[0]].extend(tmp_a)
#        R_norm_ = [R_norm_[0]].extend(tmp_r)
        [E_norm_[0]].extend(tmp_a)
        [R_norm_[0]].extend(tmp_r)

        #print '4',E_norm_, R_norm_

        E_idx = E_idx + [idx_.tolist()]
        E_norm_tmp = E_norm_tmp + [E_norm_]
        R_norm_tmp = R_norm_tmp + [R_norm_]
        #sys.stdout.flush()
        #aaa = input()

    return E_idx,  E_norm_tmp, R_norm_tmp


def make_R(L, E_norm, R_norm_tmp, R, Amax):
    Rmax = np.empty(L,dtype=int)
    for i in range(L):
        x = E_norm[i]
        y = R_norm_tmp[i]
        Rmax[i] = np.round(interpolate.pchip_interpolate(x, y, Amax)*R[-1,i])
    return Rmax


def make_R_from_unique_points(L, E_norm, E_idx, R_norm_tmp, R, Amax):
    Rmax = np.empty(L,dtype=int)
    for i in range(L):
#        R_norm_tmp[E_idx[i],i]=1
        x = E_norm[:E_idx[i]+1,i]
        y = R_norm_tmp[:E_idx[i]+1,i]

        # refinement norm
#        if y[-1] < 1:
#            x[-1] = (1 - x[-2])*0.99 + x[-2]
#            x = np.append(x,1)
#            y = np.append(y,1)

        x, x_idx = np.unique(x,return_index=True)
        y = y[x_idx]
        y, y_idx = np.unique(y,return_index=True)
        x = x[y_idx]
        Rmax[i] = np.round(interpolate.pchip_interpolate(x, y, Amax)*R[-1,i])
    return Rmax

def Refinement_complexity(C_acc, Ctarget, space, delta):
    flag_end = 1
    if C_acc <= Ctarget+space:
        delta = delta+0.001
    else :
        delta = delta-0.001
    if (C_acc < (Ctarget +space + 0.001)) & ((Ctarget +space - 0.001) < C_acc) :
        flag_end = 0

    return delta, flag_end

def Calculate_C(r_min_vbmf, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig):
    Test_R = np.copy(r_min_vbmf)
    Test_C = Test_R*c #Test_R.dot(c)

    if Sel_conv1 != 0:
        Test_C[:,0] = 0

    C_acc = np.sum(Test_C,axis=1,dtype=np.float64)
    if Sel_comp == 1:
        C_acc = (C_acc + VGG16_fc)/C_orig
    else :
        C_acc = (C_acc)/C_orig

    return C_acc


def Ro_determination(C_acc_list, a_norm, Ctarget, space, E_norm, E_idx, R_norm_tmp, R, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig, L):

    flag_end = 1
    delta = 0
    k = 0
    while (flag_end>0) :
        # layer-wise accuracy metric
        Amax = interpolate.pchip_interpolate(C_acc_list, a_norm, Ctarget+space+delta)
        if Ctarget+space+delta > 1:
            Amax = 1.0

        # Ro determination
        #Rmax = make_R_from_unique_points(L, E_norm, E_idx, R_norm_tmp, R, Amax)
        #for i in range(len(R_norm_tmp)): R_norm_tmp[i][-1] = 1
        Rmax = make_R(L, E_norm, R_norm_tmp, R, Amax)

        # complexity
        Test_C = Rmax*c
        if Sel_conv1 != 0:
            Test_C[0] = 0

        if Sel_comp == 1:
            C_acc = (np.sum(Test_C,dtype=np.float64) + VGG16_fc)/C_orig
        else :
            C_acc = (np.sum(Test_C,dtype=np.float64))/C_orig

        # refinement of complexity
        delta, flag_end = Refinement_complexity(C_acc, Ctarget, space, delta)
        k = k+1
        if k > 500 :
            break
        
    return Rmax

def ENC_Map(Ctarget,eigen_cumsum, Sel_comp, space, Sel_conv1, Sel_cost, Sel_norm):

    # losd network specification
    W_orig, C_orig, VGG16_fc, Net, W_norm, Wmax, C_norm, Cmax, R_norm, L, A, R, W, C = Net_spec(Sel_comp)

    N = np.shape(R_norm)[0]

    R = R[:,:L]
    C = C[:,:L]
    W = W[:,:L]
 
    a = Net[:,7]/Net[:,1]
    w = Net[:,6]/Net[:,1]


    # complexity setting ----------
    C_conv1 = Net[0,7]  
    c = a

    if Sel_conv1 == 1:
        C_conv1 = C_conv1/2

    if Sel_conv1 != 0:
        Ctarget = Ctarget - C_conv1/C_orig

    # make E norm ------------
    E_norm = np.empty([np.shape(R_norm)[0],L],dtype=float)
    A_norm = A.astype(float)/A[-1]

    for i in range(L):
        if Sel_norm == 0 :
            E_norm[:,i] = np.array([(eigen_cumsum[i][0][R[n,i]-1]-eigen_cumsum[i][0][0])/(eigen_cumsum[i][0][-1]-eigen_cumsum[i][0][0]) for n in range(N)])
        else :
            E_norm[:,i] = A_norm[:,i]
        E_norm[0,i] = 0

    # ====================================
    # 1. Calculate the Re
    # ====================================
    a_norm = np.arange(0.1, 1.01, 0.01)
#    R_norm_tmp = np.copy(R_norm)
    r_min_vbmf = np.empty([len(a_norm),L],dtype=int)

#    E_mask = E_norm<1
#    E_idx = np.sum(E_mask,0)

    E_idx, E_norm_tmp, R_norm_tmp = make_unique_points(L, E_norm, R_norm)
    for n in range(len(a_norm)):
        #r_min_vbmf[n] = make_R_from_unique_points(L, E_norm, E_idx, R_norm_tmp, R, a_norm[n])
        r_min_vbmf[n] = make_R(L, E_norm_tmp, R_norm_tmp, R, a_norm[n])

    # ====================================
    # 2. Calculate the complexity from Re
    # ====================================
    C_acc = Calculate_C(r_min_vbmf, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig)
    print max(C_acc), Ctarget

    #aaa = input()
    # ===
    if Ctarget > max(C_acc):
        for i in range(len(R_norm_tmp)): R_norm_tmp[i][-1] = 1
        for n in range(len(a_norm)):
            r_min_vbmf[n] = make_R(L, E_norm_tmp, R_norm_tmp, R, a_norm[n])

        C_acc = Calculate_C(r_min_vbmf, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig)


    # ====================================
    # 3. Ro Determination
    # ====================================
    a_norm = np.insert(a_norm,0,[0])
    C_acc_list = np.insert(C_acc,0,[0])
    C_acc_out = np.zeros(2)
    W_acc_out = np.zeros(2)

    C_acc_list, c_idx = np.unique(C_acc_list,return_index=True)
    a_norm = a_norm[c_idx]
    Rmax = Ro_determination(C_acc_list, a_norm, Ctarget, space, E_norm_tmp, E_idx, R_norm_tmp, R, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig, L)
    Rmin = Ro_determination(C_acc_list, a_norm, Ctarget, -space, E_norm_tmp, E_idx, R_norm_tmp, R, c, Sel_conv1, Sel_comp, VGG16_fc, C_orig, L)

    if Sel_conv1 == 1 : # conv1 - half
        Rmax[0] = round(Net[0,1]/2)
        Rmin[0] = round(Net[0,1]/2)
#        C_acc_out[0] = Rmax.dot(a)/C_orig
#        C_acc_out[1] = Rmin.dot(a)/C_orig
#        C_acc_out = C_acc_out + C_conv1/C_orig

    elif Sel_conv1 == 2 : # conv1 - max
        Rmax[0] = round(Net[0,1])
        Rmin[0] = round(Net[0,1])
#        C_acc_out[0] = Rmax.dot(a)/C_orig
#        C_acc_out[1] = Rmin.dot(a)/C_orig
#        C_acc_out = C_acc_out + C_conv1/C_orig

    C_acc_out[0] = (Rmax.dot(a))/C_orig
    C_acc_out[1] = (Rmin.dot(a))/C_orig
    W_acc_out[0] = (Rmax.dot(w))/W_orig
    W_acc_out[1] = (Rmin.dot(w))/W_orig



    print('C range : {},   W range : {}'.format(C_acc_out, W_acc_out))
    print'Initial Rmax : ',
    for i in range(len(Rmax)):
         print int(Rmax[i]),


    print'\nInitial Rmin : ',
    for i in range(len(Rmin)):
         print int(Rmin[i]),

    print('\n')


    return C_acc_out, W_acc_out, Rmax, Rmin



